src/engine/IdTable.h

```diff
@@ -13,557 +13,751 @@
+    } else {
     }
 
-    Row(Row&& other)
-        : _data(other._allocated ? other._data : new Id[other._cols]),
-          _cols(other._cols),
-          _allocated(true) {
+    return *this;
+  }
+
+  Row& operator=(Row&& other) {
+    // Check for self assignment.
+    if (&other == this) {
+      return *this;
+    }
+    // This class cannot use move semantics if at least one of the two
+    // rows invovlved in an assigment does not manage it's data, but rather
+    // functions as a view into an IdTable
+    if (_allocated) {
+      // If we manage our own storage recreate that to fit the other row
+      delete[] _data;
       if (other._allocated) {
+        // Both rows manage their own storage so we can take advantage
+        // of move semantics.
+        _data = other._data;
+        _cols = other._cols;
+
+        // otherwise the data will be deleted unexpectedly
         other._data = nullptr;
-      } else {
-        std::memcpy(_data, other._data, sizeof(Id) * _cols);
-      }
-    }
 
-    Row& operator=(const Row& other) {
-      // Check for self assignment.
-      if (&other == this) {
         return *this;
-      }
-      if (_allocated) {
-        // If we manage our own storage recreate that to fit the other row
-        delete[] _data;
+      } else {
         _data = new Id[other._cols];
         _cols = other._cols;
       }
-      // Copy over the data from the other row to this row
-      if (_cols == other._cols && _data != nullptr && other._data != nullptr) {
-        std::memcpy(_data, other._data, sizeof(Id) * _cols);
-      }
-      return *this;
     }
-
-    Row& operator=(Row&& other) {
-      // Check for self assignment.
-      if (&other == this) {
-        return *this;
-      }
-      // This class cannot use move semantics if at least one of the two
-      // rows invovlved in an assigment does not manage it's data, but rather
-      // functions as a view into an IdTable
-      if (_allocated) {
-        // If we manage our own storage recreate that to fit the other row
-        delete[] _data;
-        if (other._allocated) {
-          // Both rows manage their own storage so we can take advantage
-          // of move semantics.
-          _data = other._data;
-          _cols = other._cols;
-
-          // otherwise the data will be deleted unexpectedly
-          other._data = nullptr;
-
-          return *this;
-        } else {
-          _data = new Id[other._cols];
-          _cols = other._cols;
-        }
-      }
-      // Copy over the data from the other row to this row
-      if (_cols == other._cols && _data != nullptr && other._data != nullptr) {
-        std::memcpy(_data, other._data, sizeof(Id) * _cols);
-      }
-      return *this;
+    // Copy over the data from the other row to this row
+    if (_cols == other._cols && _data != nullptr && other._data != nullptr) {
+      std::memcpy(_data, other._data, sizeof(Id) * _cols);
```

Comment:
couldn't `oter._data` be the one we previously set `nullptr` in line 418? I.e. we should only do the `std::memcpy` if other was a view.

Reply:
No, because line 420 returns.
I just checked again (this logic has been here for quite some time)
and all 4 cases (this and other can both be allocated or views) are handled correctly.

If both are allocated, we just redirect pointers and return (line 411).
In any other case we have to do the `memcpy` either to the IdTable directly (this is a view) or to newly
allocated storage (this is allocated).